cmd/contour/serve.go

```diff
@@ -216,6 +216,36 @@ func NewServer(log logrus.FieldLogger, ctx *serveContext) (*Server, error) {
 		Scheme:                 scheme,
 		MetricsBindAddress:     "0",
 		HealthProbeBindAddress: "0",
+		NewCache: ctrl_cache.BuilderWithOptions(ctrl_cache.Options{
+			// TransformByObject is a function that allows changing incoming objects before they are cached by the informer.
+			// This is useful for saving memory by removing fields that are not needed by Contour.
+			TransformByObject: ctrl_cache.TransformByObject{
+				&corev1.Secret{}: func(obj interface{}) (interface{}, error) {
+					secret, _ := obj.(*corev1.Secret)
```

Comment:
The godoc for `TransformFunc` says:

> // TransformFunc (similarly to ResourceEventHandler functions) should be able
> // to correctly handle the tombstone of type cache.DeletedFinalStateUnknown

See e.g. https://github.com/projectcontour/contour/blob/main/internal/dag/cache.go#L279-L280 for how we handle that in the DAG cache currently - I guess we should do something similar here.

Reply:
Thanks, I had missed this! 

I added check for the success of type assertion for both `TransformByObject` and `DefaultTransform`. I just pass the original object if it fails.

In case of `TransformByObject` the function is explicitly assigned for type `corev1.Secret` but I did not find anything that suggests that it could not get passed an `cache.DeletedFinalStateUnknown` as well, so it makes sense to add this. 

Though, I'm bit confused why the [`TransformFunc`](https://github.com/kubernetes/client-go/blob/8cbca742aebe24b24f7f4e32fd999942fa9133e8/tools/cache/controller.go#L356-L365) would be passed `cache.DeletedFinalStateUnknown` in any situation, if the tombstone object is associated to [`OnDelete`](https://github.com/kubernetes/client-go/blob/8cbca742aebe24b24f7f4e32fd999942fa9133e8/tools/cache/controller.go#L209-L212)? I thought transformation would happen when new objects are introduced into the cache, not when they are removed. Maybe the transformers are called also at delete 🤨.